content/post/siva.md

```diff
@@ -0,0 +1,207 @@
+description: "When analyzing all the source code from all VCS repositories, storage becomes a major problem. śiva, our archive format, is part of the solution."
+categories: ["go", "git", "technical"]
+---
+
+## Motivation
+
+At source{d} we fetch, store and analyze millions of git repositories.
+Today we are covering ~16M and our goal is to cover all of them, as well as keep
+up with the rapid growth of code hosted online.
+
+Now, what is in a git repository? For a full explanation, you should read
+[Pro Git book, Chapter 10](https://git-scm.com/book/en/v2/Git-Internals-Plumbing-and-Porcelain),
+but for our purposes, there are only a few relevant pieces:
+
+ * `config` file with repository-specific git configuration.
+ * `objects/pack` directory containing, at least, one [packfile](https://git-scm.com/book/en/v2/Git-Internals-Packfiles)
+ and its index file.
+ * `HEAD` file with the HEAD reference.
+ * (optional) `refs` directory containing [references](https://git-scm.com/book/en/v2/Git-Internals-Git-References).
+ * (optional) `packed-refs` file with [packed references](https://git-scm.com/docs/git-pack-refs).
+
+For each repository, we store, at least, 5 files (config, 1 packfile,
+1 packfile index, HEAD reference and a master reference). This grows, at least,
+by two (1 incremental packfile and its index) every time with update a repository.
+
+For a fresh fetch of our current ~16M repository, that adds up to 80M files and
+32M more for each update. That is without considering the high monthly growth of
+repository number itself.
+
+We store all of this in a distributed file system. Currently Google Cloud
+Storage and we are evaluating HDFS for our bare metal migration. Either way,
+whenever we need to update or analyze a repository, we need to read most of the
+repository files. Neither GCS or HDFS are low-latency systems, so the latency of
+each file fetch from the distributed file system dominates the time to fetch a
+repository to a computing node.
+
+At this point, it was clear to us that we need to archive repositories. That way,
+when processing a repository, we would need to fetch a single file from the
+distributed file system.
+
+## Our Ideal Archive Format
+
+So we started evaluating different archiving formats to use. That should be easy!
+There are the features we are looking for:
+
+1. **No compression.** Most data is contained in packfiles, which are already
+  compressed.
+2. **Seekable.** We need random access to arbitrary positions of the archived
+   files.
+3. **Indexed.** Fast access to specific files.
+4. **Cheap Add.** With every update, we need to add two files
+  to the archive, so ideally, we should be able to add these files to the archive
+  without completely rewriting it.
+5. **Concatenable.** Ideally, it should be possible to add new files to the archive
+  with a single append operation. We could use [GCS compose](https://cloud.google.com/storage/docs/composite-objects)
+  or HDFS append.
```

Comment:
Added clarification when explaining packfiles int he motivation, as well as in this paragraph.